allennlp/training/trainer.py

```diff
@@ -339,25 +340,30 @@ def _restore_checkpoint(self) -> int:
     def from_params(cls,
                     model: Model,
                     serialization_dir: str,
-                    optimizer: torch.optim.Optimizer,
                     iterator: DataIterator,
                     train_dataset: Dataset,
                     validation_dataset: Optional[Dataset],
-                    params: Optional[Params] = None) -> 'Trainer':
+                    params: Params) -> 'Trainer':
 
-        params = params or Params({})
         patience = params.pop("patience", 2)
         validation_metric = params.pop("validation_metric", "-loss")
         num_epochs = params.pop("num_epochs", 20)
         cuda_device = params.pop("cuda_device", -1)
         grad_norm = params.pop("grad_norm", None)
         grad_clipping = params.pop("grad_clipping", None)
         lr_scheduler_params = params.pop("learning_rate_scheduler", None)
+
+        if cuda_device >= 0:
+            model = model.cuda(cuda_device)
```

Comment:
This is now duplicated in the init.  Would it make sense to remove it from there?  Not sure...

Reply:
Yeah, struggled a bit with that - if you aren't building from params, this would then have different behaviour, but equally, whether we want that behaviour there or not I don't know.